prometheus/vec.go

```diff
@@ -231,19 +243,115 @@ func (m *MetricVec) hashLabels(labels Labels) (uint64, error) {
 		if !ok {
 			return 0, fmt.Errorf("label name %q missing in label map", label)
 		}
-		h = hashAdd(h, val)
+		h = m.hashAdd(h, val)
 	}
 	return h, nil
 }
 
-func (m *MetricVec) getOrCreateMetric(hash uint64, labelValues ...string) Metric {
-	metric, ok := m.children[hash]
+// getOrCreateMetric retrieves the metric by hash and label value or creates it
+// and returns the new one.
+//
+// This function takes a lock, so avoid
+func (m *MetricVec) getOrCreateMetric(hash uint64, values interface{}) Metric {
+	m.mtx.RLock()
+	metric, ok := m.getMetric(hash, values)
+	m.mtx.RUnlock()
+	if ok {
+		return metric
+	}
+
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+	metric, ok = m.getMetric(hash, values)
 	if !ok {
 		// Copy labelValues. Otherwise, they would be allocated even if we don't go
 		// down this code path.
-		copiedLabelValues := append(make([]string, 0, len(labelValues)), labelValues...)
-		metric = m.newMetric(copiedLabelValues...)
-		m.children[hash] = metric
+		labelValues := m.copyLabelValues(values)
```

Comment:
The comment here is not correct anymore ("Copy labelValues..." – it's called `values` now). And perhaps move the comment into the `copyLabelValues` method because it only applies to the case where `values` is a `[]string` (otherwise it needs to be extracted out of the Labels map anyway).